torchtext/csrc/vectors.cpp

```diff
@@ -62,23 +72,32 @@ struct Vectors : torch::CustomClassHolder {
   }
 
   void __setitem__(const std::string &token, const torch::Tensor &vector) {
-    const auto &item = stovec_.find(token);
-    if (item != stovec_.end()) {
-      item->value() = vector;
+    const auto &item_index = stoindex_.find(token);
+    if (item_index != stoindex_.end()) {
+      stovec_.insert_or_assign(token, vector);
+      vectors_[item_index->value()] = vector;
     } else {
-      tokens_.push_back(token);
-      vectors_ = torch::cat({vectors_, torch::unsqueeze(vector, /*dim=*/0)},
-                            /*dim=*/0);
-      stovec_.insert_or_assign(token, vectors_.select(0, stovec_.size()));
+      stoindex_.insert_or_assign(token, vectors_.size(0));
+      stovec_.insert_or_assign(token, vector);
+      // TODO: This could be done lazily during serialization (if necessary).
+      // We would cycle through the vectors and concatenate those that aren't
+      // views.
+      vectors_ = at::cat({vectors_, vector.unsqueeze(0)});
     }
   }
 
   int64_t __len__() { return stovec_.size(); }
 };
 
 VectorsStates _set_vectors_states(const c10::intrusive_ptr<Vectors> &self) {
+  std::vector<std::string> tokens(self->stoindex_.size());
+  // reconstruct tokens list
+  for (const auto &item : self->stoindex_) {
```

Comment:
Keep in mind that you (as far as I know) don't have a guarantee on order here.

Reply:
Wait could you explain what you mean. You can get the ordering from the index that the token corresponds to. So by me placing each token in its corresponding index in the `tokens` vector, I am getting the correct order am I not?